Added support for excluding files from sync

segphault opened the pull request:
Implements the feature requested in #150.

ppolewicz commented on b2/sync.py:
```diff
@@ -775,17 +780,19 @@ def make_folder_sync_actions(source_folder, dest_folder, args, now_millis, repor
-            sync_type, source_file, dest_file, source_folder, dest_folder, args, now_millis
-        ):
-            yield action
+        if not any([is_excluded(ex, source_file) or
```
shouldn't only the source side be excluded? Also, please use a generator expression rather than a list comprehension here - there is no need to allocate a temporary list.

ppolewicz commented on b2/sync.py:
```diff
@@ -752,6 +753,8 @@ def make_file_sync_actions(
                 yield LocalDeleteAction(dest_file.name, version.id_)
 
+def is_excluded(ex, target_file):
+    return target_file is not None and ex.match(target_file.name)
```
why would target_file be None?

ppolewicz commented on b2/sync.py:
```diff
@@ -775,17 +780,19 @@ def make_folder_sync_actions(source_folder, dest_folder, args, now_millis, repor
+                    is_excluded(ex, dest_file) for ex in exclusions]):
 
+            if source_folder.folder_type() == 'local':
+                if source_file is not None:
```
why protect against `None`?

ppolewicz commented:
Travis CI tests have failed. Could you please make sure the code you submit passes tests, and that new functionality is tested as well? We rely on those tests quite a bit.

segphault commented:
When zip_files generates a union of the two trees, it will yield None in cases where the file exists on one side but not the other. Without that check in `is_excluded`, the application would throw an error complaining that it can't access the `name` property on `target_file`.

It's worth noting that in the second case that you pointed out, the code in question predates my patch: https://github.com/Backblaze/B2_Command_Line_Tool/blob/master/b2/sync.py#L779

ppolewicz commented on b2/console_tool.py:
```diff
@@ -546,7 +546,7 @@ def run(self, args):
     """
     b2 sync [--delete] [--keepDays N] [--skipNewer] [--replaceNewer] \\
-            [--threads N] [--noProgress] <source> <destination>
+            [--threads N] [--noProgress] [--exclude N] <source> <destination>
```
perhaps `[--exclude <regex>]` would be better?

bwbeach replied:
How about:  `[--exclude <regexp>]`

ppolewicz commented:
You can run `pre-commit.sh` to fix code style automatically.

ppolewicz commented:
All looks good, except the missing test.

segphault commented:
Thanks! I added a test case for exclusion. Does that look good?

bwbeach commented:
Sorry for not commenting earlier.  I was AFK yesterday.

Nice work!

How do you want `--exclude` to interact with `--delete`?  If you exclude a file from the sync, should it get deleted if it exists at the destination?  I think I would say "yes".  You could accomplish this by filtering the files from the source folder before they get zipped with the files from the destination folder.

bwbeach commented:
I was debating with myself yesterday whether regexp or glob would be better for pattern matching, and whether excluding a folder name should automatically exclude everything in it.

Regexp is more powerful, but harder to use.  I'm fine with using it.

I might like to automatically exclude files in folders, though.  I think that's a common use case, and making people stick `/.*` on the end of the regexp seems unnecessary.  As it is, saying `--exclude /my/folder` will do nothing, because the regexp won't match `/my/folder/file1.txt`.

ppolewicz commented:
```python
>>> re.match('foo/', 'foo/bar')
<_sre.SRE_Match object at 0x7fcd4d9c0370>
>>>
```
so sticking `/.*` at the end is not necessary.

As for glob vs. regex, maybe we can use `--exclude-regex` and if someone will want glob later, we can add `--exclude`?

segphault commented:
Thanks for the feedback! I moved the exclusion check so that it's performed inside of zip_folders on the `all_files` iterator. That appears to make the feature work as expected with the the --delete option. I added an extra test case to specifically check exclusion and deletion together.

I also renamed the option to --excludeRegex as @ppolewicz  suggests so that somebody else can add an --exclude option with glob support at some point later on.
